crates/bevy_ecs/src/term_query/builder.rs

```diff
@@ -0,0 +1,84 @@
+use std::marker::PhantomData;
+
+use crate::{
+    component::ComponentId,
+    prelude::{Component, With, Without, World},
+};
+
+use super::{ComponentTerm, QueryTermGroup, Term, TermQueryState, TermVec};
+
+pub struct QueryBuilder<'w, Q: QueryTermGroup = ()> {
+    terms: TermVec<Term>,
+    current_term: usize,
+    world: &'w mut World,
+    _marker: PhantomData<Q>,
+}
+
+impl<'w, Q: QueryTermGroup> QueryBuilder<'w, Q> {
+    pub fn new(world: &'w mut World) -> Self {
+        let mut terms = TermVec::new();
+        Q::init_terms(world, &mut terms);
+        Self {
+            current_term: terms.len(),
+            terms,
+            world,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    pub fn term<T: QueryTermGroup>(&mut self) -> &mut Self {
+        T::init_terms(self.world, &mut self.terms);
+        self
+    }
+
+    pub unsafe fn term_at(&mut self, index: usize) -> &mut Self {
+        self.current_term = index;
+        self
+    }
+
+    pub fn with<T: Component>(&mut self) -> &mut Self {
+        self.term::<With<T>>()
+    }
+
+    pub fn with_id(&mut self, id: ComponentId) -> &mut Self {
```

Comment:
I think these methods will be much clearer if they're called `foo_by_id`: `with_id` for example sounds like something completely different.